src/libraries/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ecc.c

```diff
@@ -24,6 +24,10 @@ int32_t AppleCryptoNative_EccGenerateKey(int32_t keySizeBits,
     {
         CFDictionaryAddValue(attributes, kSecAttrKeyType, kSecAttrKeyTypeEC);
         CFDictionaryAddValue(attributes, kSecAttrKeySizeInBits, cfKeySizeValue);
+        if (__builtin_available(macOS 10.15, iOS 13, tvOS 13, *))
```

Comment:
Where does the syntax from this come from?

Reply:
It's Apple clang extension (also available in Objective-C through more discoverable keyword). It compares the version numbers against the version of OS you are targeting. The effect is that it links some code to check for OS version if the enclosed block specified higher version than the minimum targeted one. It also changes the linking of the enclosed functions to be weak (can be checked with `nm obj.o` vs `nm -no-weak obj.o`). Essentially it's the official mechanism to target new APIs when linking for older OS versions. It's slightly more type safe and easier to use than `dlsym` (which is also a bit problematic on iOS App Store).

Reply:
https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available

Reply:
Actually I think the weak linking is only based on the availability attributes decorating the API declaration and the macOS/iOS version specified on command line; not on the present inside of the `__builtin_available` block.

Reply:
_I've rewritten my original answer to make it more clear:_

The [__builtin_available](https://clang.llvm.org/docs/LanguageExtensions.html#objective-c-available) function and [`availability` attribute](https://clang.llvm.org/docs/AttributeReference.html#availability) are Apple clang extensions that work in unison.

Apple platforms have a concept of weak linking in the object file format where the dynamic linker will try to bind symbols but instead of a hard failure if the symbol is not available it will bind `null` value instead.

The `availability` attribute is present on all the APIs in the SDK header through the `API_AVAILABLE` and `API_UNAVAILABLE` macros. When compiling against a newer SDK and specifying that you target an older OS version through `-mmacosx-version-min=10.13` option the new APIs added after macOS 10.13 will be still available but they will automatically get weakly linked. For example, the API [SecCertificateCopyKey](https://github.com/phracker/MacOSX-SDKs/blob/041600eda65c6a668f66cb7d56b7d1da3e8bcc93/MacOSX11.3.sdk/System/Library/Frameworks/Security.framework/Versions/A/Headers/SecCertificate.h#L148-L157) will get weakly linked (verifiable by checking output of `nm` vs `nm --no-weak` on the output .o or .dylib).

To make sure you don't call weakly linked APIs on platforms where the API was not available yet the other feature comes into play. The call to `__builtin_available` (or `@available` in Objective-C) is substituted by the compiler with a call that checks the OS version for you. The exact substitution depends on the minimum OS version you are targeting. If you specify `if (__builtin_available(macOS 10.12, *))` and target macOS 10.13 it becomes an `if (true)`. If you specify `if (__builtin_available(macOS 10.15, *))` and target macOS 10.13 it will turn into `if (__isPlatformVersionAtLeast(10, 15))`. The exact name and implementation of the `__isPlatformVersionAtLeast` method is irrelevant, the compiler makes sure it's statically linked in.

Finally, last part of the compiler feature is that you will get a warning (`-Wunguarded`/`-Wunguarded-new`) if you target f.ex. macOS 10.13 and try to call an API available on macOS 10.15. The warnings are suppressed within an appropriately guarded `if` block that checks OS version through `__builtin_available`.